### ISSUE TITLE
WavelengthNumberControl track and thumb cannot be customized.

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/scenery-phet#730]
In https://github.com/phetsims/beers-law-lab/issues/278 and other occurrences of WavelengthSlider, I'm trying to use WavelengthNumberControl.   I need to customize the slider track and thumb, and that's currently impossible.  You can only specify the trackHeight; other aspects are hard-coded, probably using values that are specific to the sim where this originated:

```js
    sliderOptions: {
          trackNode: new SpectrumSliderTrack( wavelengthProperty, options.range, {
            valueToColor: VisibleColor.wavelengthToColor,
            size: new Dimension2( 160, trackHeight )
          } ),
          thumbNode: new SpectrumSliderThumb( wavelengthProperty, {
            valueToColor: VisibleColor.wavelengthToColor,
            width: 25,
            height: 25,
            cursorHeight: trackHeight
          } )
```

Things that need to be fixed:

* trackNode and thumbNode should be excluded from NumberControlOptions.sliderOptions.  WavelengthNumberControl sets those, and they must be SpectrumSliderTrack and SpectrumSliderThumb instances.

* Slider options related to the default track and thumb should be excluded from NumberControlOptions.sliderOptions, see this assertion in Slider:

```
    // Guard against mutually exclusive options before defaults are filled in.
    assert && assertMutuallyExclusiveOptions( providedOptions, [ 'thumbNode' ], [
      'thumbSize', 'thumbFill', 'thumbFillHighlighted', 'thumbStroke', 'thumbLineWidth', 'thumbCenterLineStroke',
      'thumbTouchAreaXDilation', 'thumbTouchAreaYDilation', 'thumbMouseAreaXDilation', 'thumbMouseAreaYDilation'
    ] );
```
 
* Add `spectrumSliderTrackOptions` and `spectrumSliderThumbOptions` to SelfOptions. Then propagate those to SpectrumSliderTrack and SpectrumSliderThumb.

@samreid was the author of WavelengthNumberControl. Thoughts?  Anything else to be generalized?

**Comment by pixelzoom** [comment]
I'm going to take a stab at the things that identified above, so I can proceed with the conversion that I'm in the middle of.

**Comment by pixelzoom** [comment]
Summary of changes:

* Added options `spectrumSliderTrackOptions?: SpectrumSliderTrackOptions` and `spectrumSliderThumbOptions?: SpectrumSliderThumbOptions`
* Added assertions `!options.sliderOptions.trackNode` and `!options.sliderOptions.thumbNode`, because I couldn't figure out how to exclude specific properties from nested options of WavelengthNumberControlOptions.
* Despite what I said above, I think it's preferable to NOT exclude other track- & thumb-related options from `sliderOptions`, and just let Slider fail like it normally does if mutually-exclusive options are used.
* Tested in wave-interference (the only existing usage) and it looked and behaved the same.
* Replaced WavelengthSlider with WavelengthNumberControl in beers-law-lab, see https://github.com/phetsims/beers-law-lab/issues/278

@samreid please review.

**Comment by pixelzoom** [comment]
After changing beers-law-lab, it started failing CT, see https://github.com/phetsims/beers-law-lab/issues/285.   WavelengthNumberControl had not been PhET-iO instrumented yet, so I did that in the above commit.   So @samreid please add that to your review.

**Comment by samreid** [comment]
I reviewed the commits and had a few questions.

* The file now has 4 calls to `optionize`, several of them with signatures like `optionize<T,{},T>`.  I'm unfamiliar with that usage and unsure what it is supposed to mean or how it works.  Should merge be used for some of those?
* There is a TODO marked like so: `//TODO https://github.com/phetsims/scenery-phet/issues/730 it would be preferable to omit these from WavelengthNumberControlOptions`.  Would something like this work?

<details>
```diff

Index: js/WavelengthNumberControl.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/WavelengthNumberControl.ts b/js/WavelengthNumberControl.ts
--- a/js/WavelengthNumberControl.ts	(revision 8f82199803a076aee1938f67a3d51b6890fcca0f)
+++ b/js/WavelengthNumberControl.ts	(date 1649286547523)
@@ -37,7 +37,9 @@
   spectrumSliderThumbOptions?: SpectrumSliderThumbOptions;
 };
 
-export type WavelengthNumberControlOptions = SelfOptions & NumberControlOptions;
+export type WavelengthNumberControlOptions = SelfOptions & Omit<NumberControlOptions, 'sliderOptions'> & {
+  sliderOptions: Omit<NumberControlOptions['sliderOptions'], 'trackNode' | 'thumbNode'>;
+};
 
 /**
  * @param {Property.<number>} wavelengthProperty - wavelength, in nm

```
</details>

I tested behavior in Wave Interference's Diffraction screen, and it seemed good.  Everything else seems good for this review.

**Comment by pixelzoom** [comment]
> * The file now has 4 calls to `optionize`, several of them with signatures like `optionize<T,{},T>`.  I'm unfamiliar with that usage and unsure what it is supposed to mean or how it works.  Should merge be used for some of those?

That pattern is how you combine defaults to create an object that conforms to a specific options API. It's being dicussed in https://github.com/phetsims/phet-core/issues/114.   Alterntively you could use `OptionizeDefaults`, but I prefer a 2nd `optionized` call -- one way of doing things, instead of 2 different patterns, see https://github.com/phetsims/geometric-optics/issues/410.

Here's another example from geometric-optics ArrowObjectNode.ts:

```js
    const arrowNode = new ArrowNode( 0, 0, 0, 1,
      optionize<ArrowNodeOptions, {}, ArrowNodeOptions>( {}, GOConstants.ARROW_NODE_OPTIONS, {
        fill: arrowObject.fill,
        stroke: null
      } ) );
```

Here's what that same example would look like with `OptionizeDefaults`:

```js
    const arrowNode: OptionizeDefaults<{}, ArrowNodeOptions> = 
      merge( {}, GOConstants.ARROW_NODE_OPTIONS, {
        fill: arrowObject.fill,
        stroke: null
      } ) );
```

**Comment by pixelzoom** [comment]
> .... I tested behavior in Wave Interference's Diffraction screen, and it seemed good. Everything else seems good for this review.

How did you test? That patch is not providing type-safety for me.

I applied the patch that @samreid provided, then made this change to DiffractionScreenView:

<details>
<summary>DiffractionScreenView patch</summary>

```diff
Index: js/diffraction/view/DiffractionScreenView.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/diffraction/view/DiffractionScreenView.js b/js/diffraction/view/DiffractionScreenView.js
--- a/js/diffraction/view/DiffractionScreenView.js	(revision c5a7634fc2a06be2e4813eaf61b99e2b30ccfa56)
+++ b/js/diffraction/view/DiffractionScreenView.js	(date 1649288305312)
@@ -256,6 +256,9 @@
 
     const wavelengthPanel = new WaveInterferencePanel(
       new WavelengthNumberControl( model.wavelengthProperty, {
+        sliderOptions: {
+          thumbNode: new Rectangle( 0, 0, 50, 50, { fill: 'red' } )
+        },
         arrowButtonOptions: {
           touchAreaXDilation: WaveInterferenceConstants.NUMBER_CONTROL_HORIZONTAL_TOUCH_AREA_DILATION,
           touchAreaYDilation: 7
```

</details> 

I got no TS errors (in WebStorm or tsc) and the sim failed at runtime with:

```
assert.js:25 Uncaught Error: Assertion failed: WavelengthNumberControl sets thumbNode
    at window.assertions.assertFunction (assert.js:25:13)
    at new WavelengthNumberControl (WavelengthNumberControl.ts:76:17)
    at new DiffractionScreenView (DiffractionScreenView.js? [sm]:258:7)
    at DiffractionScreen.createView (DiffractionScreen.js? [sm]:37:16)
    at DiffractionScreen.initializeView (Screen.ts:282:23)
    at Array.<anonymous> (Sim.js? [sm]:785:16)
    at Sim.js? [sm]:793:11
```

**Comment by samreid** [comment]
```ts
const arrowNode = new ArrowNode( 0, 0, 0, 1,
  optionize<ArrowNodeOptions, {}, ArrowNodeOptions>( {}, GOConstants.ARROW_NODE_OPTIONS, {
    fill: arrowObject.fill,
    stroke: null
  } ) );
```

Can you help me understand why the `ParentOptions` is also set to `ArrowNodeOptions` in cases like these?

**Comment by pixelzoom** [comment]
> Can you help me understand why the ParentOptions is also set to ArrowNodeOptions in cases like these?

I had the same question, and don't like the look of this pattern. Again, see https://github.com/phetsims/phet-core/issues/114.

**Comment by pixelzoom** [comment]
```
        sliderOptions: { // should be forbidden
          thumbNode: new Rectangle( 0, 0, 50, 50, { fill: 'red' } )
        },
```

@samreid I think you intended for this comment to be one the next line down, correct? `sliderOptions` is OK, `thumbNode` is not.

**Comment by zepumph** [comment]
I think I'm generally confused by this issue

>Neither of those faulty options is caught by the type checker.

What file are you in? It looks like DiffractionScreenView.js to me, were you testing in typescript?

As for the type, are you asking for my help in how to remove thumbNode and trackNode from providedOptions? Because there is currently nothing doing that in the type system, so we just need to decide how we would like that to look. Here are two options. Option 2 doesn't work, but I feel like we could get it to with a bit more effort, and it looks a bit nicer to me.

```diff
Index: js/WavelengthNumberControl.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/WavelengthNumberControl.ts b/js/WavelengthNumberControl.ts
--- a/js/WavelengthNumberControl.ts	(revision fd0dc152dcaa7113cd0ce29cf6007f0ba3843d83)
+++ b/js/WavelengthNumberControl.ts	(date 1649351876768)
@@ -19,6 +19,8 @@
 import optionize from '../../phet-core/js/optionize.js';
 import Tandem from '../../tandem/js/Tandem.js';
 import Slider from '../../sun/js/Slider.js';
+import Property from '../../axon/js/Property.js';
+import { Rectangle } from '../../scenery/js/imports.js';
 
 const wavelengthNMValuePatternString = sceneryPhetStrings.wavelengthNMValuePattern;
 const wavelengthString = sceneryPhetStrings.wavelength;
@@ -36,8 +38,22 @@
   spectrumSliderTrackOptions?: SpectrumSliderTrackOptions;
   spectrumSliderThumbOptions?: SpectrumSliderThumbOptions;
 };
+//
+// // Option 1  // // This worked, but duplicates some code
+type NoThumbOrTrackNumberControlOptions = NumberControlOptions & {
+  sliderOptions?: {
+    thumbNode?: never;
+    trackNode?: never;
+  };
+};
+
+// Options 2, I wanted this to work, but it didn't
+// type NoThumbOrTrackNumberControlOptions = Omit<NumberControlOptions, 'sliderOptions'> & {
+//   sliderOptions?: Omit<NumberControlOptions['sliderOptions'], 'thumbNode' | 'trackNode'>;
+// }
 
-export type WavelengthNumberControlOptions = SelfOptions & NumberControlOptions;
+
+export type WavelengthNumberControlOptions = SelfOptions & NoThumbOrTrackNumberControlOptions;
 
 /**
  * @param {Property.<number>} wavelengthProperty - wavelength, in nm
@@ -106,4 +122,11 @@
   }
 }
 
+const x = new WavelengthNumberControl( new Property<number>( 0 ), {
+  sliderOptions: {
+    thumbNode: new Rectangle()
+  }
+} );
+
+
 sceneryPhet.register( 'WavelengthNumberControl', WavelengthNumberControl );
\ No newline at end of file

**Comment by pixelzoom** [comment]
> As for the type, are you asking for my help in how to remove thumbNode and trackNode from providedOptions? 

Yes. That's why I originally added this TODO in WavelengthNumberControl.  And I was aware that there's a deficiency for nested options.

```
//TODO https://github.com/phetsims/scenery-phet/issues/730 it would be preferable to omit these from WavelengthNumberControlOptions.
```

**Comment by samreid** [comment]
I think there are reasonable proposals for stripping out the trackNode and thumbNode from the options, but the type checker isn't catching it due to https://github.com/phetsims/phet-core/issues/116.

**Comment by zepumph** [comment]
Noting as part of https://github.com/phetsims/chipper/issues/1128

**Comment by samreid** [comment]
Current status:

```ts
type TestThing = {
  name: string;
  nestedOptions?: {
    height: number;
    width: number;
    age: number;
  };
};

type DeepOmit2<A, B extends keyof A, C extends keyof Required<A>[B]> = Omit<A, B> & Partial<Record<B, Omit<Required<A>[B], C>>>;

type OM = DeepOmit2<TestThing, 'nestedOptions', 'height' | 'age'>;
const m: OM = {
  name: 'person',
  nestedOptions: {
    // age: 123,
    // height: 123,
    width: 123
  }
};
```

But this isn't working correctly if `nestedOptions` is required.  We may just need a conditional checking if the mid-level nesting is optional or required.

I also saw stackoverflow notes on deep omit, but they versions I checked were not type safe (could specify nonexistent keys), but it was nice that they were arbitrarily deep. https://stackoverflow.com/questions/55539387/deep-omit-with-typescript

**Comment by samreid** [comment]
`https://github.com/microsoft/TypeScript/issues/23126` had a description of how to determine whether a type attribute is optional, and I used it to create this conditional type that seems to work properly:

```ts
type TestThing = {
  name: string;
  nestedOptions?: {
    height: number;
    width: number;
    age: number;
  };
};

type Tack<A, B extends keyof A, C extends keyof Required<A>[B]> = Record<B, Omit<Required<A>[B], C>>;

type DeepOmit2<A, B extends keyof A, C extends keyof Required<A>[B]> = Omit<A, B> & ( undefined extends A[B] ?
                                                                                      Partial<Tack<A, B, C>> :
                                                                                      Tack<A, B, C> );

type OM = DeepOmit2<TestThing, 'nestedOptions', 'height' | 'age'>;
const m: OM = {
  name: 'person',
  // nestedOptions: {
  //   age: 123,
  //   // height: 123,
  //   width: 123
  // }
};
```

With that definition, we can omit `trackNode` and `thumbNode` from the `NumberControlOptions` like so:

```ts
export type WavelengthNumberControlOptions = SelfOptions & DeepOmit2<NumberControlOptions, 'sliderOptions', 'trackNode' | 'thumbNode'>;
```

**Comment by pixelzoom** [comment]
Is `DeepOmit` really "deep"?  It looks like it's limited to the first level of nesting. If I've understood that correctly, then perhaps it should be named something like `NestedOmit`.

**Comment by pixelzoom** [comment]
And if we're looking for a general solution (which would indeed need to be "deep"), then I don't think this addresses cases like `numberControlOptions.sliderOptions.thumbNode`  in:

```
type SelfOptions = {
  numberControlOptions: NumberControlOptions;
}
```

**Comment by zepumph** [comment]
It would definitely be nice to have an arbitrarily deep type that can narrow nested options as we want to. I also feel like it would be nice to have a pleasant way to omit an option from the public API and add that option yourself. You often see this pattern:

type SelfOptions = { myNodeOptions?: NodeOptions; };

type MyClassOptions = SelfOptions;


class MyClass {

  constructor( providedOptions?: MyClassOptions ) {

    const options = optionize<MyClassOptions, SelfOptions>()( {
      myNodeOptions: {
        x: 10,
        y: 10
      }
    }, providedOptions );
    const myNode = new Node( options.myNodeOptions );
  }
}


x and y can be overridden with options if you want. That is totally fine. Next we say that we set visible, and the public API cannot:

```ts

type SelfOptions = { myNodeOptions?: Omit<NodeOptions, 'visible'>; };

type MyClassOptions = SelfOptions;


class MyClass {

  constructor( providedOptions?: MyClassOptions ) {

    const options = optionize<MyClassOptions, SelfOptions>()( {
      myNodeOptions: {
        x: 10,
        y: 10
// Cannot have visible here because it has been omitted
      }
    }, providedOptions );
    const myNode = new Node( assignOptions<NodeOptions>( {
      visible: false
    }, options.myNodeOptions ) );
  }
}
```

In this way above, we have to use a whole separate merge call because optionize won't allow `visible` within it's defaults object, but it would read much better if all nested options were in a single place, and Typescript was in charge of the public API.

```ts


type SelfOptions = { myNodeOptions?: NodeOptions; };

// Omit only from the public API
type MyClassOptions = NestedOmit<SelfOptions, 'myNodeOptions', 'visible'>;


class MyClass {

  constructor( providedOptions?: MyClassOptions ) {

    const options = optionize<MyClassOptions, SelfOptions>()( {
      myNodeOptions: {
        x: 10,
        y: 10,
        visible: false
      }
    }, providedOptions );
    
    const myNode = new Node( options.myNodeOptions );
  }
}
```

**Comment by pixelzoom** [comment]
Here's an example that I provided in Slack of the approach that I try to use for nested options, in order to keep the type definitions from getting messy.  The general idea is to add the nested options that the client is responsible for when the subcomponent that received those nested options is being instantiated, using `combineOptions`.

As @zepumph said in Slack, it's not the pattern for all situations. But give it a shot if you see your type definitions getting overly complicated.

```js
type SelfOptions = {

  // MyNode has responsibility for the button's content, listener, and tandem.
  helloButtonOptions?: Omit<RectangularPushButtonOptions, 'content' | 'listener' | 'tandem'>;
}

type MyNodeOptions = SelfOptions & NodeOptions & PickRequired<NodeOptions, 'tandem'>;

export default class MyNode extends Node {

  constructor( providedOptions: MyNodeOptions ) {

    super( providedOptions );

    const helloButton = new RectangularPushButton(
      combineOptions<RectangularPushButtonOptions>( {
        content: new Text( 'hello' ),
        listener: () => console.log( 'hello world' ),
        tandem: providedOptions.tandem.createTandem( 'helloButton' )
      }, providedOptions.helloButtonOptions )
    );
    this.addChild( helloButton );

    // add other UI components ...
  }
}
```

**Comment by samreid** [comment]
The proposal `DeepOmit2` demonstrates a type-safe way to achieve nested omits.  If we like the way it works for depth 2 and our only qualm is we need more levels, we can apply the same strategy to create `DeepOmit3` etc.  We may even be able to shield usages by `type DeepOmit = DeepOmit2 | DeepOmit3 | DeepOmit4` etc. so that clients only need to use `DeepOmit`.  I haven't tested that it but it is plausible.

As I mentioned in https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1120348122, I have seen arbitrarily deep omit implementations, some of which use dot syntax.  Like `DeepOmit<GOPanelOptions,'numberControlOptions.sliderOptions.thumbNode'>`.  But they did not check that the key (path) actually exists.  The proposal for `DeepOmit2` checks that.

**Comment by zepumph** [comment]
I think we should work toward a type with 3 arguments, the Type, the path into the nested object, and the list of keys to omit. This will be very useful in these cases.

`DeepOmit<GOPanelOptions,'numberControlOptions.sliderOptions', 'thumbNode' | 'trackNode'>

@samreid does that seem possible?

**Comment by samreid** [comment]
I'm not sure.  Some details:

* I don't know how to make the dot syntax be type safe.  Not sure if there is a way
* If you want to exclude `numberControlOptions.sliderOptions.thumbNode` and `numberControlOptions.sliderOptions.trackNode.fill` would that be separate DeepOmit calls?  As long as we're brainstorming, how about DeepOmit just takes one string that is the full path?  Or an array of full string paths?  Not advocating here, just brainstorming.

**Comment by zepumph** [comment]
To me one call per level feels fine, and likely more succinct than all those paths. I'm good with anything though. I don't know how exactly to use the dot notation, even after some research, so I don't think I can help much here.

**Comment by samreid** [comment]
Here's an adaptation of https://stackoverflow.com/questions/67785778/typescript-deep-omit-with-dot-notation which works like this:

```ts
type Nested = {
  a: number;
  b: {
    c: number;
    d: number;
  };
};
type Omitted = DeepOmit<Nested, 'b.c'>;
const foo: Omitted = { a: 1, b: { c: 1, d: 2 } }; // type error, c not allowed
```

The main problem is you can put `DeepOmit<Nested, 'b.e'>` and it won't complain.  Is that OK?

<details>

```diff

type Tail<T extends any[]> = ( ( ...t: T ) => void ) extends ( ( h: any, ...r: infer R ) => void ) ? R : never;

type DeepOmit1<T, Path extends string[]> = T extends object ?
                                           Path['length'] extends 1 ? Omit<T, Path[0]> : {
                                             [K in keyof T]: K extends Path[0] ? DeepOmit1<T[K], Tail<Path>> : T[K];
                                           } : T;

type UnDot<T extends string> =
  T extends `${infer A}.${infer B}`
  ? [ A, ...UnDot<B> ]
  : [ T ];

type Nested = {
  a: number;
  b: {
    c: number;
    d: number;
  };
};

type DeepOmit<A, B extends string> = DeepOmit1<A, UnDot<B>>;

type Omitted = DeepOmit<Nested, 'b.c'>;

const foo: Omitted = { a: 1, b: { c: 1, d: 2 } }; // error
```

</details>

**Comment by samreid** [comment]
This helped a little bit (checking if the value appeared on the T), but didn't flag at the DeepOmit site or work correctly:

```ts
type DeepOmit1<T, Path extends string[]> = T extends object ?
                                           Path['length'] extends 1 ? ( Path[0] extends keyof T ? Omit<T, Path[0]> : never ) : {
                                             [K in keyof T]: K extends Path[0] ? DeepOmit1<T[K], Tail<Path>> : T[K];
                                           } : T;
```

**Comment by zepumph** [comment]
>The main problem is you can put DeepOmit<Nested, 'b.e'> and it won't complain. Is that OK?


My guess is that this is a deal breaker. Is it not for you?

**Comment by samreid** [comment]
I'd prefer to have type safety in that part, even if it means we have to adopt a less convenient syntax at the call site.  Even something like:

```ts
DeepOmit4<MyType,'something','element','child','colorOptions' | 'fillOptions'>
```

**Comment by samreid** [comment]
I was starting to work on an implementation that uses the bracket accessor like so:

```ts
type TestThing = {
  name: string;
  nestedOptions?: {
    height: number;
    width: number;
    age: number;
  };
};

// type ArbitraryOmit<A,

type OM = Omit<TestThing, 'nestedOptions'> & { nestedOptions: Omit<Required<TestThing>['nestedOptions'], 'age'> }
```

And was surprised to find that `Omit` doesn't check its key values:

```ts
type TestThing = {
  name: string;
};

type OM = Omit<TestThing, 'anotehunatoheu'>; // not a type error
```

If we like that precedent and want to follow it, then we could use the dot operator as above.  If we think this is unfortunate, then we can stick with checking that the keys are present.  But that will narrow the possible solution paths.

**Comment by pixelzoom** [comment]
@samreid said:
> I'd also like to hear about the last few checkboxes in https://github.com/phetsims/phet-info/issues/191#issuecomment-1125632378 before spending more time here.

Response is in https://github.com/phetsims/phet-info/issues/191#issuecomment-1126256710.  We can modify the examples when we change the approach.

**Comment by samreid** [comment]
Now awaiting discussion of https://github.com/phetsims/phet-core/issues/119 before we continue.  After that, we should discuss the value/importance of a deep omit in light of https://github.com/phetsims/phet-info/issues/191#issuecomment-1125632378

**Comment by samreid** [comment]
We decided we prefer type-safe Omit in https://github.com/phetsims/phet-core/issues/119, so we would probably want type safety in DeepOmit as well.  We have also reviewed https://github.com/phetsims/phet-info/issues/191#issuecomment-1125632378 and maybe should decide on the importance of DeepOmit in that context.

**Comment by samreid** [comment]
I tested the DeepOmit implementation listed at https://github.com/ts-essentials/ts-essentials#deepomit.  It has a nice syntax for nesting the omit structure, but it is also non-strict:

```ts
interface Teacher {
  name: string;
  gender: string;
  students: { name: string; score: number }[];
}

type TeacherSimple = DeepOmit<Teacher,
  {
    gender: never;
    strategy: never; // Not marked as a type error
    students: {
      score: never;
    }[];
  }>;
```

**Comment by chrisklus** [comment]
From 6/16/22 dev meeting:

SR: There are no known type-safe ways to implement DeepOmit for arbitrarily deep nested options. We can handle as deep of cases as we'd like with DeepOmit2, DeepOmit3, etc. These are each tricky to implement. How does this sound?

SR: we can also wait on this and see if better options come up.

@pixelzoom is going to take a look at this again for status and next steps recommendations. Thanks!

**Comment by chrisklus** [comment]
JB: I haven't needed this yet
JG: Haven't needed this, but a little wary of how complicated options are getting.
CK: Haven't needed this yet either

**Comment by pixelzoom** [comment]
Current status of this issue:

* Most of the work that identified in the first comment has been done.

* This TODO in WavelengthNumberControl remains to be addressed:

```typescript
    //TODO https://github.com/phetsims/scenery-phet/issues/730 it would be preferable to omit these from WavelengthNumberControlOptions
    if ( options.sliderOptions ) {
      assert && assert( !options.sliderOptions.trackNode, 'WavelengthNumberControl sets trackNode' );
      assert && assert( !options.sliderOptions.thumbNode, 'WavelengthNumberControl sets thumbNode' );
    }
```

* @samreid was investigating how to added that TODO via something like `DeepOmit`, the limitations of which are addressed in https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1158021334.

* Other developers have not run into this problem. It could be that it will never come up elsewhere. But I think it's more likely that this kind of case is not being properly handled (either via assert or type description) when designing options APIs.

I'll reminate on this a bit, and make a recommendation on how to proceed.

**Comment by pixelzoom** [comment]
As an alternative to DeepOmit, I investigated removing `sliderOptions`, then adding it back without `trackNode` and `thumbNode`. 

If I export `ExtendedSliderOptions` from NumberControl.ts (we'd want to rename that type), then this works:

```typescript
export type WavelengthNumberControlOptions = SelfOptions &
  StrictOmit<NumberControlOptions, 'sliderOptions'> &
  StrictOmit<ExtendedSliderOptions, 'trackNode' | 'thumbNode'>;
```

@samreid what do you think? Would the above be acceptable?

**Comment by pixelzoom** [comment]
Ah, here we go...  This is immune to changes to the type of `NumberControlOptions.sliderOptions`:

```typescript
export type WavelengthNumberControlOptions = SelfOptions &
  StrictOmit<NumberControlOptions, 'sliderOptions'> &
  StrictOmit<Required<NumberControlOptions>[ 'sliderOptions'], 'trackNode' | 'thumbNode'>;
```

But now I see that this is basically what @samreid's `DeepOmit` was doing.  So...

I think we should rename `DeepOmit` to `NestedStrictOmit`, document that it only works for 1 level deep, and move on.

**Comment by pixelzoom** [comment]
Adapting @samreid's `DeepOmit2` that is shown in https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1120348122, I came up with `NestedStrictOmit`:

```typescript
// Omit properties from options that are nested one level deep.
type NestedStrictOmit<T, property extends keyof T, keys extends keyof Required<T>[property]> =
  StrictOmit<T, property> &
  Partial<Record<property, StrictOmit<Required<T>[property], keys>>>;
```

Example usage for WavelengthControl:

```typescript
type WavelengthNumberControlOptions = SelfOptions &
  NestedStrictOmit<NumberControlOptions, 'sliderOptions', 'trackNode' | 'thumbNode'>;
```

And `NestedStrictOmit` seems to be strict, which was a dificiency with `DeepOmit`.  Some examples:

```
// TS2344: Type '"foo"' does not satisfy the constraint ...
NestedStrictOmit<NumberControlOptions, 'foo', 'trackNode' | 'thumbNode'>;

// TS2344: Type '"trackNode" | "foo"' does not satisfy the constraint ...
NestedStrictOmit<NumberControlOptions, 'sliderOptions', 'trackNode' | 'foo'>;
```

**Comment by pixelzoom** [comment]
I went ahead and pushed `NestedStrictOmit` and changes to `WavelengthNumberControl`.  If/when we need to omit nested options that are deeper than 1 level, we can revisit.    @samreid please review, and let me know if this sounds like an OK appropach for now.

**Comment by zepumph** [comment]
Looks like we are in subgroups right now, please bring back to dev meeting discussion if needed.

**Comment by samreid** [comment]
I wrote this harness to test `NestedStrictOmit`:

```ts
type Person = {
  age: number;
  family: {
    size: number;
    location: string;
  };
};

const person = {
  age: 99,
  family: {
    size: 1,
    location: 'local'
  }
} as const;

type StrictPerson = NestedStrictOmit<Person, 'family', 'size'>

const m: StrictPerson = person;

console.log( m.family.location );
const v = m.family;
```

I observed a type error on the 2nd to last line:

<img width="723" alt="image" src="https://user-images.githubusercontent.com/679486/176194833-2a552397-9377-4f19-934a-ebe8a5adf546.png">

And I observed that family is coming up as something ` | undefined`:

<img width="749" alt="image" src="https://user-images.githubusercontent.com/679486/176194920-c6993a30-34cb-401e-8010-3413f9126a7c.png">

I swapped out the NestedStrictOmit implementation for the one described in https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1120493611.  NOTE: Not the one in the comment https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1120348122 mentioned by @pixelzoom in https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1163577110

I added StrictOmit instead of Omit, and it seems to have the desirable behavior:

```ts
type Omit2<A, B extends keyof A, C extends keyof Required<A>[B]> = Record<B, StrictOmit<Required<A>[B], C>>;
type NestedStrictOmit<A, B extends keyof A, C extends keyof Required<A>[B]> = StrictOmit<A, B> & ( undefined extends A[B] ?
                                                                                            Partial<Omit2<A, B, C>> :
                                                                                            Omit2<A, B, C> );

```

<img width="591" alt="image" src="https://user-images.githubusercontent.com/679486/176195825-d23dfcf9-c614-4eda-8ce2-26b4af3dc16c.png">

And I observed that non-existent keys are forbidden:

<img width="607" alt="image" src="https://user-images.githubusercontent.com/679486/176195982-13b83431-9770-49ac-905e-b2bcfc4d64fc.png">

<img width="670" alt="image" src="https://user-images.githubusercontent.com/679486/176196026-83922d5b-be4e-450c-aef8-05d40f392d61.png">

**Comment by pixelzoom** [comment]
@samreid Thanks for the testing in https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1168757149. 

From https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1163592619:
> @samreid please review, and let me know if this sounds like an OK approach for now.

So... Is this approach OK? If so, OK to close? If not, what's next?

**Comment by samreid** [comment]
> So... Is this approach OK? If so, OK to close?

No, the proposed NestedStrictOmit implementation in https://github.com/phetsims/phet-core/commit/760e1dadde280840fd26d25c526c0f30c9128aee returns `undefined` when it should not.  `m.family` should be defined but it is flagged as potentially undefined.

The counterproposal is for an implementation like so:

```ts
type Omit2<A, B extends keyof A, C extends keyof Required<A>[B]> = Record<B, StrictOmit<Required<A>[B], C>>;
type NestedStrictOmitVersion2<A, B extends keyof A, C extends keyof Required<A>[B]> = StrictOmit<A, B> & ( undefined extends A[B] ?
                                                                                            Partial<Omit2<A, B, C>> :
                                                                                            Omit2<A, B, C> );
```

which appears to behave properly in this test harness.  We would rename it to `NestedStrictOmit` but I wanted to give it a unique name for this comment for purposes of discussion.  Can you please review/test that implementation and see if it is OK?

**Comment by pixelzoom** [comment]
Sigh... I would be really nice if developers got in the habit of either using meaningful variable names for generics, or at least documenting meaningless variables like A/B/C.

I will review.

**Comment by pixelzoom** [comment]
The revised implementation of `NestedStrictOmit` works great.  I tested with the examples in https://github.com/phetsims/scenery-phet/issues/730#issuecomment-1168757149, and with a usage of `new WavelengthNumberControl`.  

This addressed this GitHub issue (nesting 1 level deep), but does not address nesting that is arbitrarily deep.  Assigning back to @samreid answers to these questions:

(1) Is this approach sufficient for `WavelengthNumberControl`?
(2) Would you like to create a new issue for the general case?
(3) OK to close this issue?

**Comment by samreid** [comment]
Looks good, let's close this issue.  Let's wait on a side "arbitrarily deep" issue until it comes up again in a future context.  My hope is by then we will have more TS expertise and will be able to implement it recursively.
